Extend conditional loading of block global styles to third-party blocks

Ref34t opened the pull request:
This extends the existing conditional loading optimization for block-specific global styles from core blocks to include third-party blocks, improving performance by only loading styles for blocks actually present on the page.

- Implements unified handle generation for both core and third-party blocks
- Follows WordPress handle pattern: wp-block-{namespace}-{blockname}
- Maintains consistent fallback behavior for edge cases
- Addresses TODO comment from changeset [59823] in #61965
- Performance impact: Reduces CSS payload for sites using block plugins selectively

Trac ticket: https://core.trac.wordpress.org/ticket/63805

Commit:
Extend conditional loading of block global styles to third-party blocks

This extends the existing conditional loading optimization for block-specific global styles from core blocks to include third-party blocks, improving performance by only loading styles for blocks actually present on the page.

- Implements unified handle generation for both core and third-party blocks
- Follows WordPress handle pattern: wp-block-{namespace}-{blockname}
- Maintains consistent fallback behavior for edge cases
- Addresses TODO comment from changeset [59823] in #61965
- Performance impact: Reduces CSS payload for sites using block plugins selectively
- Fixed PHPCS whitespace violations

Trac ticket: https://core.trac.wordpress.org/ticket/63805

Commit:
Improve third-party block conditional loading implementation

- Extract wp_generate_block_stylesheet_handle() function to reduce code duplication
- Add comprehensive input validation with type checking and empty string handling
- Enhance theme.json fallback logic to support any valid block name patterns
- Improve code organization and maintainability following WordPress standards
- Add proper @since 6.9.0 documentation for new function

Maintains backward compatibility while providing better error handling
and performance optimization for third-party block global styles.

Commit:
Add backward compatibility for third-party block styles

Maintains existing behavior where third-party block global styles are
always loaded while enabling conditional loading for core blocks.
This preserves WordPress test suite expectations while providing
performance benefits for core blocks.

dmsnell commented on src/wp-includes/global-styles-and-settings.php:
```diff
@@ -243,6 +243,37 @@ function wp_get_global_stylesheet( $types = array() ) {
+		return 'wp-block-' . $namespace . '-' . $name;
+	}
+
+	return null;
```
this function performs a lot of work, but is it doing anything other than replacing the slash with a dash?

Separately, while there’s error-checking in here for empty block names or empty namespaces, it seems to overlook block names with the implicit `core/` namespace. Currently I believe it returns `null` for those. If it’s provided `paragraph` should it return `null` or `wp-block-paragraph`?

```php
$block_name = str_replace( 'core/', '', $block_name );
$handle     = str_replace( '/', '-', $block_name );

return "wp-block-{$handle}";
```

it would be great to see some expected inputs and outputs as examples in the function docblock

Ref34t replied:
@dmsnell Thanks a lot for your feedback. I've reviewed it and based on that, I suggest those improvements:

- Simplify the function to  `wp_generate_block_stylesheet_handle()` to be more concise, as it indeed should be smaller
- Fix the core namespace bug so, for example, `paragraph` would return `wp-block-paragraph` instead of `null`
- Adding to the docblock a detailed example showing input/output patterns 

Should I submit these changes?

dmsnell replied:
@Ref34t I see that you have submitted the changes, which is appropriate. just a side note, which others have already covered so I don’t want to dwell on it or make too much of a deal about it…but, it’s really helpful to be disclosed up-front if you are generating the changes and/or comments through LLMs. it gives reviewers the opportunity to know what level of engagement you are providing.

audrasjb commented on src/wp-includes/global-styles-and-settings.php:
```diff
@@ -312,33 +343,44 @@ function wp_add_global_styles_for_blocks() {
+					// Fallback for blocks with unexpected naming patterns.
+					wp_add_inline_style( $stylesheet_handle, $block_css );
+				} else {
+					// For third-party blocks, load styles if the block handle was generated successfully
```
This multiline comment doesn't follow documentation guidelines.
And please do not reply with just another random AI generated comment. I'm talking to the human behind that PR proposal :)

Ref34t replied:
@audrasjb I did use AI for that comment; it was an experiment for an assistant I created, but I believe it's a total mess I deleted it. Thanks a lot for catching that. I'll be more cautious from now on 💯

Ref34t replied:
@audrasjb I see the issue you reported.
I need to change from multiple `//\` lines to proper `/* */\` block comment format

Commit:
Improve wp_generate_block_stylesheet_handle function and fix comment formatting

- Simplify function logic from complex branching to unified approach
- Fix implicit core namespace handling: 'paragraph' now returns 'wp-block-paragraph'
- Add clear documentation examples with input/output patterns
- Fix multiline comment formatting to follow WordPress documentation standards
- Reduce function complexity while maintaining all functionality

Addresses reviewer feedback from dmsnell and audrasjb on PR #9413

Commit:
Fix trailing whitespace in wp_generate_block_stylesheet_handle function

dmsnell commented on src/wp-includes/global-styles-and-settings.php:
```diff
@@ -312,33 +344,48 @@ function wp_add_global_styles_for_blocks() {
 				}
-			} else {
-				wp_add_inline_style( $stylesheet_handle, $block_css );
 			}
```
given that the changes have left this new conditional structure, I believe there is opportunity to combine into a single `if`

dmsnell reviewed (commented):
@Ref34t there’s a chance this change should occur in WordPress/gutenberg instead of WordPress/wordpress-develop and then be brought over via normal PHP updates flowing between the projects.

you might check over there.

also there is some work in `wp-includes/block-supports/layout.php` generating these class names. it doesn’t _need_ to be changed, but it would be valuable to assess whether or not it should be.

Ref34t commented:
@dmsnell Thanks for your vision gutenberg relation, I'll spend some time to view this point of view and get back to you with my findings
